### PR TITLE
Geoprocessing API: Add Auth Tokens

### PR DESCRIPTION
**Pull request by arottersman**
## Overview

Secure the geoprocessing API with DRF auth tokens. This will allow us:
- to better track requests
- to throttle particular users
- to stop troublemakers by revoking their tokens

Connects #2187 

### Demo

screen shot 2017-09-11 at 5 36 33 pm

### Notes

* We're currently not using the api token at all on the frontend
* Downgraded `django-rest-swagger` to account for marcgibbons/django-rest-swagger#497

### 1. Tokens created?
 * After pulling this branch, `./scripts/manage.py migrate`
       - this will create the `authtoken` table and add tokens for your existing users
 * For an existing user, call
```http
http POST http://localhost:8000/api/api-token-auth/ username="***" password="***"
```
Confirm the response is something like
```json
{ "token" : "9944b09199c62bcf9418ad846dd0e4bbdfc6ee4b" }
```
- Create a new user in the app and repeat the step above to confirm a token was created for it

### 2. App still works?
- Confirm you can still use the app as usual, logged in or not

### 3. Tokens doing their job?
- Using either of the two tokens you got in step (1)....
  - go to `localhost:8000/api/docs/`
  - enter the token into the upper right and confirm you can use the endpoints like before
  - refresh and don't enter the token. You should only get 401 responses
  - enter a garbage token. You should get 401s.
 - Use a real token in the "Authorization" header parameter field for the requests, and confirm there are results

- Change [this line](https://github.com/WikiWatershed/model-my-watershed/compare/arr/api-authtoken?expand=1#diff-607ea8eb5dc05eb093a0a2971e42e8eeR23) to 
```python
whitelisted = false
```
- Run the following requests

_Should be 200_
```http
http POST 'localhost:8000/api/analyze/land/?wkaoi=huc12__55174' "Authorization: Token <your_token>"
```
_Should be 401_
```http
http POST 'localhost:8000/api/analyze/land/?wkaoi=huc12__55174'
```

### CODE REVIEW

**Comment by mmcfarland** [comment]
For reference, the testing URL should be 
`http POST http://localhost:8000/api/api-token-auth/ username="***" pasword="***"`

File: src/mmw/mmw/settings/base.py [hunk 1]
```diff
@@ -326,6 +326,7 @@ def get_env_setting(setting):
         'licenseUrl': 'http://www.apache.org/licenses/LICENSE-2.0.html',
         'title': 'Model My Watershed API',
     },
+    'is_authenticated': True,
```
[the remark this answers is not in the document]
Comment (arottersman) [reply src/mmw/mmw/settings/base.py]: Not my intention. Will remove!

File: src/mmw/apps/modeling/views.py [hunk 2]
```diff
@@ -482,7 +485,9 @@ def drb_point_sources(request):
 
 
 @decorators.api_view(['GET'])
-@decorators.permission_classes((AllowAny, ))
+@decorators.authentication_classes((SessionAuthentication,
```
[the remark this answers is not in the document]
Comment (arottersman) [reply src/mmw/apps/modeling/views.py]: I'll look into this. I added the `SessionAuthentication` so the `request.user` would get set on requests from the client.

**Comment by arottersman** [comment]
This is ready for another look.

* Removed swagger setting that required you to be logged into the app
* For `/jobs`, put `TokenAuthentication` before `SessionAuthentication` so the token takes precedence. This should fix the issue where you couldn't get jobs via swagger while signed into the app.

**Comment by arottersman** [comment]
Thanks for your review!

**Comment by arottersman** [comment]
A little baffled as to why the build fails:

```
  File "/opt/app/apps/geoprocessing_api/urls.py", line 11, in <module>
    from apps.geoprocessing_api import views
  File "/opt/app/apps/geoprocessing_api/views.py", line 823, in <module>
    @decorators.permission_classes((AllowAny, ))
NameError: name 'AllowAny' is not defined
```

But that geoprocessing view line doesn't exist anymore.

On my local, `check` and `test` still pass.

**Comment by arottersman** [comment]
Thank goodness for jenkins. There was a sneaky addition of the analyze climate endpoint on develop that was causing the build to fail. Added a the same `TokenAuthentication` permissions, and a swagger docstring.

**Comment by mmcfarland** [comment]
Tested out the climate endpoint and the bigcz app and am getting:

```
AttributeError at /api/analyze/climate/
'TokenAuthentication' object has no attribute 'has_permission'
```

Other endpoints and app functions are still working, including the `job` endpoint in swagger now.

**Comment by arottersman** [comment]
That's unfortunate. Any beef if I were to install a module like [this](http://nginx.org/en/docs/http/ngx_http_realip_module.html) to pass along the original IP as a header?

**Comment by arottersman** [comment]
> Tested out the climate endpoint and the bigcz app and am getting ...
Added 31e229b to fix.

**Comment by mmcfarland** [comment]
> an nginx module

I don't believe that would alleviate the problem.  Requests for both "app" and api clients would terminate SSL the ELB, proxy through nginx and arrive at the view with the same IP.  If it were always the actual client IP, it would still be indistinguishable since app requests originate from a client browser.

I think the options we have are:
a) Give each environment for the app an api key, and risk it being used by other accounts
b) Use an app key like above, but insist that usage comes from a valid domain via `HTTP_ORIGIN` or `HTTP_REFERER`.  This could be spoofed by a client, but would be accurate for browsers
c) Set a flag via the cookie, similar to (or actually) CSRF tokens. 

On other sites which provide a key meant for client mapping (like Google Maps), it seems common to pair the publicly exposed key with a whitelisted domain.

What do you think of any of those options, or others?

**Comment by arottersman** [comment]
Oh. Hm, I guess that should have been obvious. I'll open a card. 

Seems like the advantage of the CSRF tokens would be that we can expire them more frequently, and (if we did use actual csrf tokens) they're already available and don't need additional maintenance. A client API token, on the other hand, would fit better in the API ecosystem we're building. Not sure which criteria are best to optimize for in this case.

Maybe the important thing is we [throttle all anonymous requests](http://www.django-rest-framework.org/api-guide/throttling/#how-clients-are-identified) via IP at a rate that someone clicking through the UI will likely never hit, but some hydrologist trying to be sneaky would?

**Comment by arottersman** [comment]
I've squashed the fixup, and am going to merge. Thanks for your help with this!

**Pull request merged**